core/src/HIP/Kokkos_HIP_UniqueToken.hpp

```diff
@@ -81,49 +83,68 @@ class UniqueToken<HIP, UniqueTokenScope::Global> {
 
   /// \brief upper bound for acquired values, i.e. 0 <= value < size()
   KOKKOS_INLINE_FUNCTION
-  size_type size() const noexcept { return m_count; }
+  size_type size() const noexcept { return m_locks.extent(0); }
 
   /// \brief acquire value such that 0 <= value < size()
   KOKKOS_INLINE_FUNCTION
   size_type acquire() const {
-    const Kokkos::pair<int, int> result =
-        Kokkos::Impl::concurrent_bitset::acquire_bounded(
-            m_buffer, m_count, Kokkos::Impl::clock_tic() % m_count);
-
-    if (result.first < 0) {
-      Kokkos::abort(
-          "UniqueToken<HIP> failure to acquire tokens, no tokens available");
-    }
-
-    return result.first;
+    KOKKOS_IF_ON_DEVICE(
+        int idx = blockIdx.x * (blockDim.x * blockDim.y) +
+                  threadIdx.y * blockDim.x + threadIdx.x;
+        idx = idx % size(); unsigned int active = __ballot(1);
+        unsigned int done_active = 0; bool done = false;
+        while (active != done_active) {
+          if (!done) {
+            if (Kokkos::atomic_compare_exchange(&m_locks(idx), 0, 1) == 0) {
+              done = true;
+            } else {
+              idx += blockDim.y * blockDim.x + 1;
+              idx = idx % size();
+            }
+          }
+          done_active = __ballot(done ? 1 : 0);
+        } return idx;
+
+// Make sure that all writes in the previous lock owner are visible to me
+#ifdef KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
```

Comment:
The compiler does not like the macro inside another macro
`/var/jenkins/workspace/Kokkos/core/src/HIP/Kokkos_HIP_UniqueToken.hpp:109:2: error: embedding a directive within macro arguments has undefined behavior [-Werror,-Wembedded-directive]
`

Reply:
Use an `impl_acquire` function like you did in Cuda.

Reply:
done